File: Week_02/G20190343020171/1.两数之和.go
package G20190343020171

/*
 * @lc app=leetcode.cn id=1 lang=golang
 *
 * [1] 两数之和
 *
 * https://leetcode-cn.com/problems/two-sum/description/
 *
 * algorithms
 * Easy (47.07%)
 * Likes:    7154
 * Dislikes: 0
 * Total Accepted:    725.8K
 * Total Submissions: 1.5M
 * Testcase Example:  '[2,7,11,15]\n9'
 *
 * 给定一个整数数组 nums 和一个目标值 target，请你在该数组中找出和为目标值的那 两个 整数，并返回他们的数组下标。
 *
 * 你可以假设每种输入只会对应一个答案。但是，你不能重复利用这个数组中同样的元素。
 *
 * 示例:
 *
 * 给定 nums = [2, 7, 11, 15], target = 9
 *
 * 因为 nums[0] + nums[1] = 2 + 7 = 9
 * 所以返回 [0, 1]
 *
 *
 */

// @lc code=start
func twoSum(nums []int, target int) []int {
	// 暴力法, 时间复杂度O(n2), cost 52ms
	// for i := 0; i < len(nums); i++ {
	// 	for j := 0; j < len(nums); j++ {
	// 		if (nums[i]+nums[j]) == target && i != j {
	// 			return []int{i, j}
	// 		}
	// 	}
	// }

	// hash code, 时间复杂度 O(n),cost time 4ms
	// table := make(map[int]int)
	// for idx := range nums {
	// 	table[target-nums[idx]] = idx
	// }
	// for idx := range nums {
	// 	if val, ok := table[nums[idx]]; ok {
	// 		if idx != val {
	// 			return []int{idx, val}
	// 		}
	// 	}
	// }

	// 一遍hash, cost 4ms, bad memory
	table := make(map[int]int)
	table[target-nums[0]] = 0
	for idx := range nums {
		if val, ok := table[nums[idx]]; ok {
			if val != idx {
				return []int{val,idx}
			}
		}
		table[target-nums[idx]] = idx
	}
	return nil
}

// @lc code=end

File: Week_02/G20190343020145/1.两数之和.java
//一遍hash表 时间复杂度O(n),我们把包含有 nn 个元素的列表遍历两次。由于哈希表将查找时间缩短到 O(1)O(1) ，所以时间复杂度为 O(n)O(n)。
//空间复杂度O(n) 所需的额外空间取决于哈希表中存储的元素数量，该表中存储了 n个元素。
// class Solution {
//     public int[] twoSum(int[] nums, int target) {
//         Map<Integer,Integer> hm = new HashMap<>(); 
//         for(int i = 0;i< nums.length;i ++){
//             hm.put(nums[i], i);
//         }  
//         for(int i = 0;i< nums.length;i ++) {
//             int findnum = target - nums[i];
//             if(hm.containsKey(findnum) && hm.get(findnum) != i)
//                 return new int[] { i, hm.get(findnum) };
//         }
//         throw new IllegalArgumentException("No two sum solution");
//     }

// }


//一遍哈希表
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer,Integer> hm = new HashMap<>(); 
        for(int i = 0;i< nums.length;i ++) {
            int findnum = target - nums[i];
            if(hm.containsKey(findnum) && hm.get(findnum) != i)
                return new int[] { i, hm.get(findnum) };
            hm.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");
    }
}